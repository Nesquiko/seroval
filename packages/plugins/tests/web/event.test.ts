import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serialize,
  serializeAsync,
  toCrossJSON,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSON,
  toJSONAsync,
} from 'seroval';
import { describe, expect, it } from 'vitest';
import EventPlugin from '../../web/event';

const EXAMPLE_EVENT_TYPE = 'example';
const EXAMPLE = new Event(EXAMPLE_EVENT_TYPE);

describe('Event', () => {
  describe('serialize', () => {
    it('supports Event', () => {
      const result = serialize(EXAMPLE, {
        plugins: [EventPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE>(result);
      expect(back).toBeInstanceOf(Event);
    });
  });
  describe('serializeAsync', () => {
    it('supports Event', async () => {
      const result = await serializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [EventPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<typeof EXAMPLE>>(result);
      expect(back).toBeInstanceOf(Event);
    });
  });
  describe('toJSON', () => {
    it('supports Event', () => {
      const result = toJSON(EXAMPLE, {
        plugins: [EventPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE>(result, {
        plugins: [EventPlugin],
      });
      expect(back).toBeInstanceOf(Event);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Event', async () => {
      const result = await toJSONAsync(Promise.resolve(EXAMPLE), {
        plugins: [EventPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<typeof EXAMPLE>>(result, {
        plugins: [EventPlugin],
      });
      expect(back).toBeInstanceOf(Event);
    });
  });
  describe('crossSerialize', () => {
    it('supports Event', () => {
      const result = crossSerialize(EXAMPLE, {
        plugins: [EventPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports Event', () => {
        const result = crossSerialize(EXAMPLE, {
          plugins: [EventPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports Event', async () => {
      const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [EventPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports Event', async () => {
        const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
          plugins: [EventPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports Event', async () =>
      new Promise<void>((resolve, reject) => {
        crossSerializeStream(Promise.resolve(EXAMPLE), {
          plugins: [EventPlugin],
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    describe('scoped', () => {
      it('supports Event', async () =>
        new Promise<void>((resolve, reject) => {
          crossSerializeStream(Promise.resolve(EXAMPLE), {
            plugins: [EventPlugin],
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
    });
  });
  describe('toCrossJSON', () => {
    it('supports Event', () => {
      const result = toCrossJSON(EXAMPLE, {
        plugins: [EventPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE>(result, {
        plugins: [EventPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(Event);
    });
  });
  describe('toCrossJSONAsync', () => {
    it('supports Event', async () => {
      const example = Promise.resolve(EXAMPLE);
      const result = await toCrossJSONAsync(example, {
        plugins: [EventPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromCrossJSON<typeof example>(result, {
        plugins: [EventPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(Event);
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports Event', async () =>
      new Promise<void>((resolve, reject) => {
        toCrossJSONStream(Promise.resolve(EXAMPLE), {
          plugins: [EventPlugin],
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
  });
});
