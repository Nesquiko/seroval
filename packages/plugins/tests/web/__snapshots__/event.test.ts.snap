// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Event > crossSerialize > scoped > supports Event 1`] = `"($R=>$R[0]=new Event("example",$R[1]={bubbles:!1,cancelable:!1,composed:!1}))($R["example"])"`;

exports[`Event > crossSerialize > supports Event 1`] = `"$R[0]=new Event("example",$R[1]={bubbles:!1,cancelable:!1,composed:!1})"`;

exports[`Event > crossSerializeAsync > scoped > supports Event 1`] = `"($R=>$R[0]=Promise.resolve($R[1]=new Event("example",$R[2]={bubbles:!1,cancelable:!1,composed:!1})))($R["example"])"`;

exports[`Event > crossSerializeAsync > supports Event 1`] = `"$R[0]=Promise.resolve($R[1]=new Event("example",$R[2]={bubbles:!1,cancelable:!1,composed:!1}))"`;

exports[`Event > crossSerializeStream > scoped > supports Event 1`] = `"($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R["example"])"`;

exports[`Event > crossSerializeStream > scoped > supports Event 2`] = `"($R=>($R[4]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new Event("example",$R[3]={bubbles:!1,cancelable:!1,composed:!1})))($R["example"])"`;

exports[`Event > crossSerializeStream > supports Event 1`] = `"$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))()"`;

exports[`Event > crossSerializeStream > supports Event 2`] = `"($R[4]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new Event("example",$R[3]={bubbles:!1,cancelable:!1,composed:!1}))"`;

exports[`Event > serialize > supports Event 1`] = `"new Event("example",{bubbles:!1,cancelable:!1,composed:!1})"`;

exports[`Event > serializeAsync > supports Event 1`] = `"Promise.resolve(new Event("example",{bubbles:!1,cancelable:!1,composed:!1}))"`;

exports[`Event > toCrossJSON > supports Event 1`] = `"{"t":40,"i":0,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":1,"p":{"k":["bubbles","cancelable","composed"],"v":[{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":3},"o":0}},"c":"seroval-plugins/web/Event"}"`;

exports[`Event > toCrossJSONAsync > supports Event 1`] = `"{"t":12,"i":0,"s":1,"f":{"t":40,"i":1,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":2,"p":{"k":["bubbles","cancelable","composed"],"v":[{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":3},"o":0}},"c":"seroval-plugins/web/Event"}}"`;

exports[`Event > toCrossJSONStream > supports Event 1`] = `"{"t":28,"i":0,"f":{"t":41,"i":1,"s":2}}"`;

exports[`Event > toCrossJSONStream > supports Event 2`] = `"{"t":29,"i":0,"a":[{"t":41,"i":4,"s":3},{"t":40,"i":2,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":3,"p":{"k":["bubbles","cancelable","composed"],"v":[{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":3},"o":0}},"c":"seroval-plugins/web/Event"}]}"`;

exports[`Event > toJSON > supports Event 1`] = `"{"t":{"t":40,"i":0,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":1,"p":{"k":["bubbles","cancelable","composed"],"v":[{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":3},"o":0}},"c":"seroval-plugins/web/Event"},"f":16383,"m":[]}"`;

exports[`Event > toJSONAsync > supports Event 1`] = `"{"t":{"t":12,"i":0,"s":1,"f":{"t":40,"i":1,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":2,"p":{"k":["bubbles","cancelable","composed"],"v":[{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":3},"o":0}},"c":"seroval-plugins/web/Event"}},"f":16383,"m":[]}"`;
